File: Python/src/Board.py
# Board.py

from dataclasses import dataclass, field
from datatypes import Piece, Colour, Squares
from Move import Move
from typing import ClassVar
from bitboard import *

def GET_RANK(sq): return (sq) // 8
def GET_FILE(sq): return (sq) % 8
def FR2SQ(f, r): return (r) * 8 + (f)

START_POS = "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1"

@dataclass
class UndoBox:
	move: Move = Move()
	castle_perms: int = 0
	enpas: ClassVar[Squares] = Squares.NO_SQ
	fifty_move: int = 0
	
class Board:
	pieces: list[Piece]  # Square -> Piece
	bitboards: list[Bitboard]
	occupancies: list[Bitboard]

	piece_num: list[int]
	king_sq: list[Squares]
	side: Colour
	enpas: Squares
	castle_perms: int
	fifty_move: int  # Counter for 50-move rule

	ply: int
	his_ply: int
	move_history: list[UndoBox]

	def __init__(self):
		self.reset_board()
	
	def reset_board(self):
		self.pieces = [Piece.EMPTY] * 64
		self.bitboards = [0 for _ in range(13)]
		self.occupancies = [0 for _ in range(3)]
		self.piece_num = [0] * 13
		self.king_sq = [Squares.NO_SQ] * 3
		self.side = Colour.WHITE
		self.enpas = Squares.NO_SQ
		self.castle_perms = 0
		self.fifty_move = 0
		self.ply = 0
		self.his_ply = 0
		self.move_history = [UndoBox() for _ in range(2048)]

	def update_vars(self):
		from datatypes import piece_type, piece_col, PType
		for pce in Piece:
			self.piece_num[pce.value] = count_bits(self.bitboards[pce.value])
			if piece_type[pce] == PType.KING:
				copy = self.bitboards[pce.value]
				sq, _ = pop_ls1b(copy)
				self.king_sq[piece_col[pce].value] = Squares(sq)

	def parse_fen(self, fen: str):
		from datatypes import Piece, piece_col
		self.reset_board()
		pfen = 0
		rank = 0
		file = 0
		piece = 0
		count = 0
		sq = 0
		fen = fen.strip()

		# Parse pieces
		while rank < 8 and pfen < len(fen):
			count = 1
			c = fen[pfen]
			if c == 'p':   piece = Piece.bP
			elif c == 'r': piece = Piece.bR
			elif c == 'n': piece = Piece.bN
			elif c == 'b': piece = Piece.bB
			elif c == 'k': piece = Piece.bK
			elif c == 'q': piece = Piece.bQ
			elif c == 'P': piece = Piece.wP
			elif c == 'R': piece = Piece.wR
			elif c == 'N': piece = Piece.wN
			elif c == 'B': piece = Piece.wB
			elif c == 'K': piece = Piece.wK
			elif c == 'Q': piece = Piece.wQ
			elif c in '12345678':
				piece = Piece.EMPTY
				count = int(c)
			elif c == '/':
				rank += 1
				file = 0
				pfen += 1
				continue
			elif c == ' ':
				pfen += 1
				break
			else:
				raise ValueError(f"Invalid FEN character: {c}")
			for _ in range(count):
				sq = rank * 8 + file
				if piece != Piece.EMPTY:
					self.pieces[sq] = piece
					self.bitboards[piece.value] = set_bit(self.bitboards[piece.value], Squares(sq))
					self.occupancies[Colour.BOTH.value] = set_bit(self.occupancies[Colour.BOTH.value], Squares(sq))
					self.occupancies[piece_col[piece].value] = set_bit(self.occupancies[piece_col[piece].value], Squares(sq))
				file += 1
			pfen += 1
		# Side to move
		if pfen < len(fen):
			self.side = Colour.WHITE if fen[pfen] == 'w' else Colour.BLACK
			pfen += 2
		# Castling rights
		for _ in range(4):
			if pfen >= len(fen) or fen[pfen] == ' ':
				break
			c = fen[pfen]
			if c == 'K': self.castle_perms |= 1
			elif c == 'Q': self.castle_perms |= 2
			elif c == 'k': self.castle_perms |= 4
			elif c == 'q': self.castle_perms |= 8
			pfen += 1
		if pfen < len(fen) and fen[pfen] == ' ':
			pfen += 1
		# En passant
		if pfen < len(fen) and fen[pfen] != '-':
			file = ord(fen[pfen]) - ord('a')
			rank = 8 - int(fen[pfen + 1])
			self.enpas = Squares(rank * 8 + file)
			pfen += 3
		else:
			self.enpas = Squares.NO_SQ
			pfen += 2

		# Fifty-move counter
		half_moves = ''
		while pfen < len(fen) and fen[pfen].isdigit():
			half_moves += fen[pfen]
			pfen += 1
		self.fifty_move = int(half_moves) if half_moves else 0
		if pfen < len(fen) and fen[pfen] == ' ':
			pfen += 1

		# Full move number (not stored)
		full_move = ''
		while pfen < len(fen) and fen[pfen].isdigit():
			full_move += fen[pfen]
			pfen += 1
		self.update_vars()


	def print_board(self):
		from datatypes import ascii_pieces
		print()
		for rank in range(8):
			print(f"  {8 - rank} ", end='')
			for file in range(8):
				sq = rank * 8 + file
				piece = self.pieces[sq]
				print(f" {ascii_pieces[piece.value]}", end='')
			print()
		print("\n     a b c d e f g h\n")
		print(f"           Side: {'White' if self.side == 0 else 'Black'}")
		print(f"     En passant: {self.enpas if self.enpas is not None else 'N/A'}")
		print(f"50-move counter: {self.fifty_move}")
		print(f"            Ply: {self.ply}")
		print(f"    History ply: {self.his_ply}")
		# Print castling rights
		cr = ''
		cr += 'K' if self.castle_perms & 1 else '-'
		cr += 'Q' if self.castle_perms & 2 else '-'
		cr += 'k' if self.castle_perms & 4 else '-'
		cr += 'q' if self.castle_perms & 8 else '-'
		print(f"       Castling: {cr}")

File: Cpp/src/Board.cpp
// Board.cpp

#include "movegen.hpp"
#include "Board.hpp"
#include "bitboard.hpp"
#include "moveio.hpp"

#include <iostream>
#include <cstdlib> // atoi()

/*
	Macro board manipulation
*/

void reset_board(Board* pos) {

	for (int sq = 0; sq < 64; ++sq) {
		pos->pieces[sq] = EMPTY;
	}

	for (int i = 0; i < 13; ++i) {
		pos->bitboards[i] = 0ULL;
		pos->piece_num[i] = 0;
		if (i < 3) {
			pos->occupancies[i] = 0ULL;
			pos->king_sq[i] = NO_SQ;
		}
	}

	pos->side = WHITE;
	pos->enpas = NO_SQ;
	pos->castle_perms = 0;
	pos->fifty_move = 0;
	pos->ply = 0;
	pos->his_ply = 0;

	for (UndoBox& box : pos->move_history) {
		box.castle_perms = 0;
		box.enpas = 0;
		box.fifty_move = 0;
		box.hash_key = 0ULL;
		box.move = 0; // NO_MOVE
	}
}

// Update other variables of the board, for when only bitboards and occupancies were setup
void update_vars(Board* pos) {
	for (int pce = wP; pce <= bK; ++pce) {
		pos->piece_num[pce] = count_bits(pos->bitboards[pce]);
		if (piece_type[pce] == KING) {
			Bitboard copy = pos->bitboards[pce];
			pos->king_sq[piece_col[pce]] = pop_ls1b(copy);
		}
	}
}

// Rewritten from VICE parse_fen() function by Richard Allbert
void parse_fen(Board* pos, const std::string FEN) {

	if (FEN.length() <= 0) {
		std::cerr << "Board parse_fen() error: Invalid FEN length.\n";
	}

	reset_board(pos);
	int pfen = 0; // Pointer to FEN character.

	/********************
	**  Parsing Pieces **
	****************** */

	int rank  = RANK_8;
	int file  = FILE_A;
	int piece = 0;
	int count = 0; // no. of consecutive empty squares / placeholder
	int sq    = 0;

	while ((rank <= RANK_1) && pfen < (int)FEN.length()) {
		count = 1;

		switch (FEN[pfen]) {
			case 'p': piece = bP; break;
			case 'r': piece = bR; break;
			case 'n': piece = bN; break;
			case 'b': piece = bB; break;
			case 'k': piece = bK; break;
			case 'q': piece = bQ; break;
			case 'P': piece = wP; break;
			case 'R': piece = wR; break;
			case 'N': piece = wN; break;
			case 'B': piece = wB; break;
			case 'K': piece = wK; break;
			case 'Q': piece = wQ; break;

			case '1':
			case '2':
			case '3':
			case '4':
			case '5':
			case '6':
			case '7':
			case '8':
				piece = EMPTY;
				count = FEN[pfen] - '0'; // number of consecutive empty squares
				break;

			case '/':
			case ' ':
				rank++;
				file = FILE_A;
				pfen++;
				continue;

			default:
				std::cerr << "Board parse_fen() error: Invalid FEN character: " << FEN[pfen] << "\n";
				return;
		}

		// Putting pieces on the board
		for (int i = 0; i < count; i++) {
			sq = FR2SQ(file, rank);
			// Skips a file if empty square
			if (piece != EMPTY) { 
				pos->pieces[sq] = piece;
				SET_BIT(pos->bitboards[piece], sq);
				SET_BIT(pos->occupancies[BOTH], sq);
				SET_BIT(pos->occupancies[piece_col[piece]], sq);
			}
			file++;
		}
		pfen++;
	}

	/********************
	* Parsing Misc Data *
	****************** */

	// Side-to-move parsing
	pos->side = (FEN[pfen] == 'w') ? WHITE : BLACK;
	pfen += 2;

	// Castling perm parsing
	for (int i = 0; i < 4; i++) {
		if (FEN[pfen] == ' ') {
			break;
		}
		switch (FEN[pfen]) {
			case 'K': pos->castle_perms |= WKCA; break;
			case 'Q': pos->castle_perms |= WQCA; break;
			case 'k': pos->castle_perms |= BKCA; break;
			case 'q': pos->castle_perms |= BQCA; break;
			default: break;
		}
		pfen++;
	}
	pfen++;

	// En passant parsing
	if (FEN[pfen] != '-') {
		file = FEN[pfen]     - 'a';
		rank = 8 - int(FEN[pfen + 1] - '1');
		pos->enpas = FR2SQ(file, rank);
		pfen += 3;
	}
	else {
		pfen += 2;
	}

	// Fifty-move counter parsing
	uint16_t half_moves = 0;
	while (pfen < (int)FEN.length() && FEN[pfen] != ' ') {
		half_moves = half_moves * 10 + (FEN[pfen] - '0');
		pfen++;
	}
	pos->fifty_move = half_moves;
	pfen++; // Move past the space

	// Full move number parsing
	uint16_t full_move = 0;
	while (pfen < (int)FEN.length()) {
		if (FEN[pfen] == ' ') {
			break;
		}
		full_move = full_move * 10 + (FEN[pfen] - '0');
		pfen++;
	}

	update_vars(pos);
}

/*
	Output functions (debug)
*/

void print_board(const Board* pos) {
    std::cout << "\n";

    for (int rank = RANK_8; rank <= RANK_1; rank++) {
        for (int file = FILE_A; file <= FILE_H; file++) {
            if (file == 0)
                std::cout << "  " << (8 - rank) << " "; // Print rank number

			uint8_t sq = FR2SQ(file, rank);
            int piece = pos->pieces[sq];
            std::cout << " " << ascii_pieces[piece];
        }

        std::cout << "\n";
    }

    std::cout << "\n     a b c d e f g h\n\n";
    std::cout << "           Side: " << (pos->side == 0 ? "White" : "Black") << "\n";
    std::cout << "     En passant: " << (pos->enpas != NO_SQ ? ascii_squares[pos->enpas] : "N/A") << "\n";
	std::cout << "50-move counter: " << (int)pos->fifty_move << "\n";
	std::cout << "            Ply: " << (int)pos->ply << "\n";
	std::cout << "    History ply: " << (int)pos->his_ply << "\n";

    // Print castling rights
    std::cout << "       Castling: "
        << ((pos->castle_perms & WKCA) ? 'K' : '-')
        << ((pos->castle_perms & WQCA) ? 'Q' : '-')
        << ((pos->castle_perms & BKCA) ? 'k' : '-')
        << ((pos->castle_perms & BQCA) ? 'q' : '-')
        << "\n";
}